clippy_lints/src/needless_bool.rs

```diff
@@ -193,9 +209,17 @@ fn check_comparison<'a, 'tcx>(
 ) {
     use self::Expression::*;
 
-    if let ExprKind::Binary(_, ref left_side, ref right_side) = e.kind {
+    if let ExprKind::Binary(op, ref left_side, ref right_side) = e.kind {
         let (l_ty, r_ty) = (cx.tables.expr_ty(left_side), cx.tables.expr_ty(right_side));
         if l_ty.is_bool() && r_ty.is_bool() {
+            if_chain! {
+                if let BinOpKind::Eq = op.node;
+                if one_side_is_unary_not(&left_side, &right_side);
+                then {
+                    span_lint_and_help(cx, BOOL_COMPARISON, e.span, "Here comes", "the suggestion");
```

Comment:
When building a suggestion, you have to understand how spans work. A lint with a suggestion has 2 spans: 1 span (L) for the lint message and 1 span (S) for the suggestion. When running `cargo fix` on a suggestion like this, the span (S) is completely removed and replaced with the code inside the `sugg` parameter of the `span_lint_and_sugg` function. The output with differing spans may look like this:

```
error: Lint message
  --> $DIR/bool_comparison.rs:119:8
   |
LL |     if a == !b {};
   |        ^^^^^^^
   |
   = help: replace it with:
   |
LL |     if a != b {}
   |        ^^^^^^

```

As you may have noticed, the `span_lint_and_sugg` function only has one `span` argument. This is because most of the time, the span of te lint message and of the suggestion are the same. This is also the case here. The output of this case is most of the time more compact and looks nicer:

```
error: Here comes
  --> $DIR/bool_comparison.rs:119:8
   |
LL |     if a == !b {};
   |        ^^^^^^^ help: replace it with: `a != b`
```
The span of the thing you want to replace is `e.span`. To build the suggestion, you can use `utils::snippet_with_applicability`. Let's name some things:

```
a == !b
^^^^^^^ e
^       left_side
     ^^ right_side
      ^ ?
```

So you have to get the span of `?`. This is the span of the `unop` expression in the `is_unary_not` function above. Building the complete suggestion, you have to do:

```rust
format!(
    "{} != {}",
    snippet_with_applicability(cx, left_side.span, &mut applicability),
    snippet_with_applicability(cx, ?.span, &mut applicability),
)
```

This won't be enough though, since you have to also build the suggestion for the symmetric case. But this is homework for you. :wink:

Reply:
Thanks for the thorough explanation. I need some time to figure things out :)

Reply:
OK,
At last, I have some spare time to work on the issue and... mess-up the PR in the process ;-)

Anyway, I think I got it and the message now looks like the one below. I will clean-up the code and submit a new PR for review soon:

```
 error: This comparison might be written more concisely
   --> $DIR/bool_comparison.rs:119:8
    |
 LL |     if a == !b {};
    |        ^^^^^^^ help: try simplifying it as shown: `a != b`
 
 error: This comparison might be written more concisely
   --> $DIR/bool_comparison.rs:120:8
    |
 LL |     if !a == b {};
    |        ^^^^^^^ help: try simplifying it as shown: `a != b`
```